crates/scmi/src/scmi.rs

```diff
@@ -0,0 +1,1380 @@
+                ParameterType::_SignedInt32 => MessageValue::Signed(i32::from_le_bytes(slice)),
+                ParameterType::UnsignedInt32 => MessageValue::Unsigned(u32::from_le_bytes(slice)),
+            };
+            debug!("SCMI parameter {}: {:?}", n, v);
+            values.push(v);
+        }
+        request.parameters = Some(values);
+    }
+
+    fn number_of_protocols(&self) -> u32 {
+        let n: usize = self.handlers.keys().unique_by(|k| k.0).count();
+        n.try_into()
+            .expect("Impossibly large number of SCMI protocols")
+    }
+
+    pub fn register_device(&mut self, device: Box<dyn ScmiDevice>) {
+        self.devices.insert(device);
+    }
+
+    fn handle_device(
+        &self,
+        device_index: usize,
+        protocol_id: ProtocolId,
+        message_id: MessageId,
+        parameters: &[MessageValue],
+    ) -> DeviceResult {
+        self.devices
+            .handle(device_index, protocol_id, message_id, parameters)
+    }
+
+    fn device_response(&self, result: DeviceResult, device_index: usize) -> Response {
+        match result {
+            Ok(mut values) => Response::from(&mut values),
+            Err(error) => match error {
+                ScmiDeviceError::NoSuchDevice
+                | ScmiDeviceError::NotEnabled
+                | ScmiDeviceError::InvalidParameters => {
+                    info!("Invalid device access: {}, {}", device_index, error);
+                    Response::from(ReturnStatus::InvalidParameters)
+                }
+                ScmiDeviceError::UnsupportedRequest => {
+                    info!("Unsupported request for {}", device_index);
+                    Response::from(ReturnStatus::NotSupported)
+                }
+                ScmiDeviceError::GenericError => {
+                    warn!("Device error in {}", device_index);
+                    Response::from(ReturnStatus::GenericError)
+                }
+            },
+        }
+    }
+
+    fn handle_device_response(&self, request: &ScmiRequest, parameters: &[usize]) -> Response {
+        let device_index = request.get_usize(0);
+        let protocol_id = request.protocol_id;
+        let message_id = request.message_id;
+        let parameter_values: Vec<MessageValue> = parameters
+            .iter()
+            .map(|i| MessageValue::Unsigned(request.get_unsigned(*i)))
+            .collect();
+        let result = self.handle_device(
+            device_index,
+            protocol_id,
+            message_id,
+            parameter_values.as_slice(),
+        );
+        self.device_response(result, device_index)
+    }
+}
+
+/// Defines handlers for SCMI protocols.
+///
+/// The function creates a [HandlerMap] and then adds SCMI message handlers to
+/// it using [HandlerMap::bind] function.  This is the place (i.e. the
+/// functions called from here) where to add bindings for SCMI protocols and
+/// messages.
+fn make_handlers() -> HandlerMap {
```

Comment:
Any reasons why is this not a constructor for HandlerMap directly?

Reply:
I like having it structured as it is.

Reply:
I agree that it doesn't make sense to have it a standalone function, unless it's a module/library export that quickly setups sensible defaults for something, which isn't the case here.

Reply:
I still think that inlining the function reduces clarity a bit for nothing good; but not a big deal, done.

Reply:
Well think of it as "namespacing" the function, essentially. It's a class function equivalent.

Reply:
So would you suggest to move all the `make_*handler*` code inside `impl`, perhaps of `HandlerMap`?

Reply:
That would make more sense to me, yes

Reply:
OK, done.